### PR TITLE
Implement anyhow context for certificate chain

### PR DESCRIPTION
**Pull request by dlachaumepalo**
## Content
This PR includes an update of `certificate_chain` module in order to implement `anyhow` context.

## Pre-submit checklist

- Branch
  - [ ] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [X] Commit sequence broadly makes sense
  - [X] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Issue(s)
Relates to #798

### CODE REVIEW

File: mithril-common/src/certificate_chain/certificate_genesis.rs [hunk 1]
```diff
@@ -14,21 +14,9 @@ use crate::{
...
-/// [CertificateGenesisProducer] related errors.
-#[derive(Error, Debug)]
-pub enum CertificateGenesisProducerError {
-    /// Error raised when a Codec error occurs
-    #[error("codec error: '{0:?}'")]
-    Codec(StdError),
-
-    /// Error raised when there is no genesis signer available
-    #[error("missing genesis signer error")]
-    MissingGenesisSigner(),
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_genesis.rs:29]: You should keep this `MissingGenesisSigner` variant

File: mithril-common/src/certificate_chain/certificate_verifier.rs [hunk 2]
```diff
@@ -27,18 +27,6 @@ pub enum CertificateVerifierError {
...
-    #[error("codec genesis error: '{0}'")]
-    CodecGenesis(String),
-
-    /// Error raised when a CertificateRetriever tries to retrieve a [Certificate].
-    #[error("certificate retriever error: '{0}'")]
-    CertificateRetriever(#[from] CertificateRetrieverError),
-
     /// Error raised when the Genesis Signature stored in a [Certificate] is invalid.
     #[error("certificate genesis error: '{0}'")]
     CertificateGenesis(#[from] ProtocolGenesisError),
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_verifier.rs:32]: This should not be needed anymore and should be replaced by a anyhow context.

File: mithril-common/src/certificate_chain/certificate_verifier.rs [hunk 3]
```diff
@@ -192,8 +183,7 @@ impl MithrilCertificateVerifier {
                     "avk to string conversion error for certificate: `{}`",
                     certificate.hash
                 )
-            })
-            .map_err(CertificateVerifierError::Codec)?;
+            })?;
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_verifier.rs:186]: You should add a anyhow context here

File: mithril-common/src/certificate_chain/certificate_verifier.rs [hunk 4]
```diff
@@ -203,8 +193,7 @@ impl MithrilCertificateVerifier {
                     "avk to string conversion error for previous certificate: `{}`",
                     certificate.hash
                 )
-            })
-            .map_err(CertificateVerifierError::Codec)?;
+            })?;
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_verifier.rs:196]: Same as above

**Review by jpraynaud — CHANGES_REQUESTED** [review]
Few adjustments are needed :slightly_smiling_face:

File: mithril-aggregator/src/database/provider/certificate.rs [hunk 5]
```diff
@@ -628,9 +629,18 @@ impl CertificateRetriever for CertificateRepository {
...
+                    &format!(
+                        "Certificate repository can not retrieve certificate: '{}'",
+                        certificate_hash
+                    ),
+                    anyhow!(e),
+                )
+            })?
+            .ok_or(CertificateRetrieverError::new(
+                &format!("Certificate does not exist: '{}'", certificate_hash),
+                anyhow!("Certificate does not exist"),
```
Comment (ghubertpalo) [inline mithril-aggregator/src/database/provider/certificate.rs:643]: Maybe a simpler approach to this error wrapper would be to create a type that just wrap an anyhow error and add contexts to the inner error. This way it is easier to iterate over the error's different contexts IMHO.

```rust
type MyError(AnyhowError);

let a = that_fails().map_err(|e| MyError(e.with_context("my error message"))?;
```

**Review by ghubertpalo — APPROVED** [review]

File: mithril-common/src/certificate_chain/certificate_retriever.rs [hunk 6]
```diff
@@ -1,18 +1,20 @@
...
-    /// Error raised when a [CertificateRetriever] tries to retrieve a [Certificate]
-    #[error("general error: '{0}'")]
-    General(String),
+#[derive(Debug)]
+pub struct CertificateRetrieverError(pub StdError);
+
+impl From<anyhow::Error> for CertificateRetrieverError {
+    fn from(error: anyhow::Error) -> Self {
+        CertificateRetrieverError(error)
+    }
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_retriever.rs:17]: You forgot these ones:
```suggestion
impl From<StdError> for CertificateRetrieverError {
    fn from(error: StdError) -> Self {
        CertificateRetrieverError(error)
    }
```

File: mithril-common/src/certificate_chain/certificate_verifier.rs [hunk 7]
```diff
@@ -178,10 +166,14 @@ impl MithrilCertificateVerifier {
         )?;
         let previous_certificate = certificate_retriever
             .get_certificate_details(&certificate.previous_hash)
-            .await?;
+            .await
+            .map_err(|e| e.0)
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_verifier.rs:170]: Maybe this is possible?
```suggestion
            .map_err(|e| e.into())
```

File: mithril-common/src/certificate_chain/certificate_verifier.rs [hunk 8]
```diff
@@ -249,16 +241,19 @@ impl CertificateVerifier for MithrilCertificateVerifier {
...
-            genesis_certificate.signed_message.as_bytes(),
-            genesis_signature,
-        )?;
+        genesis_verifier
+            .verify(
+                genesis_certificate.signed_message.as_bytes(),
+                genesis_signature,
+            )
+            .map_err(|e| anyhow!(e))
+            .with_context(|| "Can not verify genesis certificate")?;
```
Comment (jpraynaud) [inline mithril-common/src/certificate_chain/certificate_verifier.rs:256]:
```suggestion
            .with_context(|| "Certificate verifier failed verifying a genesis certificate")?;
```

**Review by jpraynaud — CHANGES_REQUESTED** [review]
Last comments, and after we're good to go :slightly_smiling_face:

File: mithril-common/src/certificate_chain/certificate_verifier.rs (hunk 7 above)
Comment (dlachaumepalo) [reply mithril-common/src/certificate_chain/certificate_verifier.rs:170]: I've already tried it but doesn't seem possible

image

Comment (dlachaumepalo) [reply mithril-common/src/certificate_chain/certificate_verifier.rs:170]: Is it better if i had this conversion:

image

and:
image

**Review by jpraynaud — APPROVED** [review]
LGTM :+1:

**Pull request merged**